LoRaWAN: Custom FSB selection in US and Australian regions

hasnainvirk opened the pull request:
### Description

We have introduced an Mbed config based method to configure a custom frequency sub-band in US915 and AU915 PHYs.  US915Hybrid PHY is removed as it was an attempt to make a similar effort. It only allowed one sub-band as it was hard coded. Now user can choose any set of sub-band or sub-bands.


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [X] Refactor
    [ ] Target update
    [ ] Feature
    [ ] Breaking change

Commit:
Adding FSB selection feature for US915 PHY

User can now define a custom frequency sub-band for the US915 PHY.
FSB_Usage.txt defines how this parameter will be configured.

Commit:
Enabling FSB support in AU915 PHY

Just like US915 PHY, user can define a custom FSB mask for AU915 PHY.
This helps deployments where base stations do not portray full feature
channel sets and choose to stick with sub-bands.

Commit:
Removing US915Hybrid PHY

This phy implementation was just to support a single sub-band in US region.
As we have decided to make FSBs configurable, we do not need this class anymore.

hasnainvirk commented:
@AnttiKauppila @kivaisan @kjbracey-arm  Please review.

kivaisan reviewed (approved):
Looks ok.

hasnainvirk commented:
@kjbracey-arm Please review.

kjbracey commented on features/lorawan/lorastack/phy/LoRaPHYAU915.cpp:
```diff
@@ -580,6 +578,8 @@ lorawan_status_t LoRaPHYAU915::set_next_channel(channel_selection_params_t* next
             // fall back to 500 kHz default channels
             current_channel_mask[4] = channel_mask[4];
+        } else {
+            current_channel_mask[4] &= channel_mask[4];
```
Why this addition, which doesn't mirror the code for 125kHz channels above? Is it because of the lack of `&=` in `link_ADR_request`? (I recall previous discussion of that, but don't remember the explanation, and there's no comment in the code).

kjbracey reviewed (commented):
Seems fine, just still unclear on the difference in mask handling between the two channel types.

hasnainvirk commented:
@kjbracey-arm Please review again. Suggestions entertained.

johanstokking commented on features/lorawan/FSB_Usage.txt:
```diff
@@ -0,0 +1,35 @@
+Frequency sub-bands in US915/US915Hybrid/AU915:
+
+ US915/US915Hybrid/AU915 PHYs define channel structures which can support up-to 72 channels for upstream.
```
`US915Hybrid` is gone now

johanstokking reviewed (commented):
Looks good to me, two comments;

- I think it would be good to clarify that this library targets PHY 1.0.2 rev B
- I would recommend writing in the documentation to use the FSB mask only on private networks where you know beforehand what FSB is being used and/or in case of ABP. With OTAA, certainly on public networks, it's highly recommended to use the channel probing strategy as defined in PHY

Commit:
Adding helpers for Channel mask manipulation

A few protected member functions are introduced in LoRaPHY class
that help manipulate channel masks in various ways.

hasnainvirk commented:
@cmonr need CI here again. Breakage doesn't relate to this PR.

hasnainvirk commented:
@kjbracey-arm Please review.

cmonr commented:
@hasnainvirk Thanks for the poke. Thought I had gotten them all earlier today.

hasnainvirk commented:
@kjbracey-arm Can you please review ?

0xc0170 commented:
/morph build

cmonr commented:
Build issue is known. Will relaunch once corrected.

cmonr commented:
/morph build

cmonr commented:
Weird.
/morph build

hasnainvirk commented:
@cmonr @0xc0170 Exporter build failed for some reason. Can you please check ?

0xc0170 commented:
Pipeline locked error, restarting

/morph export-build

cmonr commented:
Pausing Test CI until 5.9.6 PR is merged.
Will restart CI shortly after.

cmonr commented:
/morph test

cmonr commented:
/morph export-build

0xc0170 commented:
This needs rebase ! I am aborting the test run, will restart

hasnainvirk commented:
@0xc0170 Rebased and locally tested once more.

0xc0170 commented:
/morph build

cmonr commented:
Pipes be closin to soon...

/morph export-build

AnttiKauppila commented:
@0xc0170 This can be merged
